code/localization/localize.cpp

```diff
@@ -724,199 +717,321 @@ void lcl_fred_replace_stuff(SCP_string &text)
-		// if we're not already using the default, then switch to our explicit default
-		lookup_map = &Lcl_ext_str_explicit_default;
-	}
+		// copy id
+		char xstr_id_buf[PARSE_ID_BUF_SIZE];
+		strncpy(xstr_id_buf, id_start, id_length);
+		xstr_id_buf[id_length] = '\0';
 
-	// get the string if it exists
-	if (lookup_map->find(str_id) != lookup_map->end()) {
-		auto lookup_result = (*lookup_map)[str_id];
+		//
+		// now we have the information we want
+		//
 
-		// copy to the outgoing string
-		if ( strlen(lookup_result) > max_len && !Lcl_unexpected_tstring_check )
-			error_display(0, "Token too long: [%s].  Length = " SIZE_T_ARG ".  Max is " SIZE_T_ARG ".\n", lookup_result, strlen(lookup_result), max_len);
+		xstr_str = text_str;
+		xstr_id = atoi(xstr_id_buf);
+		if (is_negative)
+			xstr_id *= -1;
+		xstr_valid = true;
 
-		strncpy(out, lookup_result, max_len);
-	}
-	// otherwise use what we have - probably should Int3() or assert here
-	else {
-		if ( strlen(text_str) > max_len && !Lcl_unexpected_tstring_check )
-			error_display(0, "Token too long: [%s].  Length = " SIZE_T_ARG ".  Max is " SIZE_T_ARG ".\n", text_str, strlen(text_str), max_len);
+		// if the localization file is not open, or there's no entry, or we're not translating, return the original string
+		if (!Xstr_inited || (xstr_id < 0) || (!use_default_translation && ((Lcl_current_lang == LCL_UNTRANSLATED) || (Lcl_current_lang == LCL_RETAIL_HYBRID))))
+			break;
 
-		strncpy(out, text_str, max_len);
-	}
+		//
+		// we are translating
+		//
 
-	// set the id #
-	if (id != NULL) {
-		*id = str_id;
-	}
+		auto lookup_map = &Lcl_ext_str;
+		if (use_default_translation && lcl_get_current_lang_index() != LCL_DEFAULT)
+		{
+			// if we're not already using the default, then switch to our explicit default
+			lookup_map = &Lcl_ext_str_explicit_default;
+		}
+
+		// get the string if it exists
+		if (lookup_map->find(xstr_id) != lookup_map->end())
+		{
+			xstr_str = (*lookup_map)[xstr_id];
+		}
+		// otherwise use what we have, but complain about it
+		else
+		{
+			mprintf(("Could not find entry %d in the external string table!\n", xstr_id));
+		}
+	} while (false);
+
+
+	// set whatever id we have
+	if (id != nullptr)
+		*id = xstr_id;
+
+	// if we made an attempt but failed, let the modder know
+	if (xstr_id == -2 && attempted_xstr)
+		error_display(0, "Malformed XSTR detected:\n\n%s\n", in);
+
+	// copy the entire string (or as much as we can)
+	strncpy(out, xstr_str, max_len);
+
+	// maybe warn about length
+	auto str_len = strlen(xstr_str);
+	if (str_len > max_len && !Lcl_unexpected_tstring_check)
+		error_display(0, "Token too long: [%s].  Length = " SIZE_T_ARG ".  Max is " SIZE_T_ARG ".\n", xstr_str, str_len, max_len);
 
-	return true;
+	return xstr_valid;
 }
 
 // ditto for SCP_string
 bool lcl_ext_localize_sub(const SCP_string &in, SCP_string &text_str, SCP_string &out, int *id, bool use_default_translation = false)
```

Comment:
It might make the code simpler if this function just called into the version above that uses C-strings, but maybe that isn't practical. I assume we wouldn't want the reverse, i.e. have that version just call into this one, because of the performance hit from creating/destroying temporary `SCP_string` objects.

If we knew what value to pass for `max_len` and could preallocate that amount of space in `text_str`/`out`, would calling the C-string version work? Or maybe a bad idea?

Reply:
I considered combining the functions, but I want to leave them as they are.  Ideally, the SCP_string version would use proper SCP_string accessor functions, but that can wait for a future upgrade.

Reply:
Okay.